var/spack/repos/builtin/packages/openmpi/package.py

```diff
@@ -526,6 +526,7 @@ class Openmpi(AutotoolsPackage, CudaPackage):
     variant(
         "orterunprefix",
         default=False,
+        when="@:4",
         description="Prefix Open MPI to PATH and LD_LIBRARY_PATH on local and remote hosts",
     )
```

Comment:
If there is a desire to optionally offer the old behavior, we could

* Keep this variant
* Use`config_args.append("--enable-prte-prefix-by-default")` For `+orterunprefix` (not essential, but explicit)
* Use `--disable-...` (if available) fort `~orterunprefix`

I get the impression though that no one was depending on `~orterunprefix`, since it sounds like we would have been effectively ignoring that for `@5:`.

It would be nice (but not strictly required) to edit logic that looks at `+orterunprefix`, like

```
        if spec.satisfies("@1.3:"):
            if spec.satisfies("+orterunprefix"):
```

(even if you don't change that, the code will still work, but it would be easier to read if for example the range were updated to `@1.3:4`)

It sounds like you were going to make additional changes based on https://github.com/spack/spack/pull/43961#issuecomment-2091092380, so let me know when that's ready and I can take another look.

Reply:
Thanks for the feedback @scheibelp! I'll be looping back to this next week and will try to take some of this into account. I agree that for completion we should maybe offer the older behavior in 5.x versions (the ability to do `~`, not currently present), even if the default behavior is changing (to `+`).

You're right that I'd still like to make some changes wrt the `romio` variant, but am having some trouble. I'll copy here a note [that I had put in Slack](https://spackpm.slack.com/archives/C5VL7V81G/p1714672523407419?thread_ts=1714606515.517369&cid=C5VL7V81G), and I'd appreciate your expert guidance:

> As written now,

    
    variant("romio", default=True, when="@:5", description="Enable ROMIO support")
    variant("romio", default=False, when="@5:", description="Enable ROMIO support")
    
    
> , it's always false. I tried adjusting this to have the True case be `@:4` instead [from our discussion above], but then I realized this is the [overriding variants](https://spack.readthedocs.io/en/latest/packaging_guide.html#overriding-variants) situation described in the docs and it continues to be False regardless of version.
To switch the +/~ behavior of a variant based on version, does it need to be separately handled or can that be done entirely as variant definitions?

Reply:
Thanks for the pointer, and yeah apologies it looks like we don't provide a way to do this with `variant` declarations. The only way I can think to do this right now is to add some config to `etc/spack/defaults/packages.yaml`:

```
packages:
  openmpi:
    prefer:
    - spec: ~romio
      when: @5:
```

(I'll check internally and get back to you on whether we'd want that sort of change, since we try to keep that scope as concise as possible)

I swear we were talking about adding a `prefer` directive (so we could encode this in the package itself), but that's not available right now. @alalazo was that you who mentioned this?

Reply:
Yes. My take is that we don't want to have `prefer` directive in `package.py`, since packages might conflict with each other (and there's no opt-out to a directive). `requires` is different, cause it's really enforcing something.

Reply:
> My take is that we don't want to have prefer directive in package.py

Fine by me, but I'm not following your reasoning

> since packages might conflict with each other (and there's no opt-out to a directive

I assume conflicting preferences would have a winner (in terms of weighting) and because the user encoded them as a preference, they'd be fine with any particular one being ignored.

Reply:
See https://github.com/spack/spack/pull/44425